File: Baekjoon/BinarySearch/[2352]반도체 설계/solution.js
const fs = require("fs");
// const inputData = fs.readFileSync("example.dat").toString().split("\n");
const inputData = fs.readFileSync("/dev/stdin").toString().split("\n");
const N = Number(inputData[0]);
const seq = inputData[1].split(" ").map((item) => Number(item));
let lis = [];

const lower_bound = (value) => {
  low = 0;
  high = lis.length;

  while (low < high) {
    mid = parseInt((low + high) / 2);

    if (value < lis[mid]) {
      high = mid;
    } else {
      low = mid + 1;
    }
  }

  return high;
};

for (let i = 0; i < N; i++) {
  idx = lower_bound(seq[i]);

  if (idx == lis.length) {
    lis.push(seq[i]);
  } else {
    lis[idx] = seq[i];
  }
}

console.log(lis.length);

File: Baekjoon/BinarySearch/[2352]반도체 설계/solution.py
import sys

N = int(sys.stdin.readline())
inputData = list(map(int, sys.stdin.readline().split()))
lis = list()


def lower_bound(value):
    low = 0
    high = len(lis)

    while low < high:
        mid = (low + high) // 2

        if value < lis[mid]:
            high = mid
        else:
            low = mid + 1

    return high


for i in range(N):
    idx = lower_bound(inputData[i])

    if idx == len(lis):
        lis.append(inputData[i])
    else:
        lis[idx] = inputData[i]

sys.stdout.write(str(len(lis)) + "\n")
